### PR TITLE
test: Notifications Inbox Screen Unit Tests Implementation

### PR DESCRIPTION
**Pull request by shafqat-muneer**
[LEARNER-10560](https://2u-internal.atlassian.net/browse/LEARNER-10560): Notifications Module Unit Tests: Inbox Screen

Adds test cases to improve coverage and ensure stability for the Notification Inbox Screen. These tests validate expected behaviors and edge cases across different scenarios.

### CODE REVIEW

File: Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+                    threadId: "5d49c25584452a0795000386",
+                    commentId: "677b2ffa7febe50470799585",
+                    postTitle: "How to learn it online?"
+                ),
+                content: "Test notification One",
+                courseId: "course-v1:edX+Test+2T2009",
+                lastRead: nil,
+                lastSeen: Date(iso8601: "2025-01-06T01:20:58.919612Z"),
+                created: Date(iso8601: "2025-01-06T01:20:58.919612Z")
+            ), SingleNotification(
```
Comment (rnr) [inline Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift]: better to start new Notification with new line

File: Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+        mockInteractor = NotificationsInteractorProtocolMock()
+        mockConnectivity = ConnectivityProtocolMock()
+        
+        Given(mockInteractor, .getAllNotifications(page: 1, willReturn: notifications))
+        
+        viewModel = NotificationsInboxViewModel(
+            notificationsInteractor: mockInteractor,
+            analytics: NotificationsAnalyticsMock(),
+            router: NotificationsRouterMock(),
+            connectivity: Connectivity(),
```
Comment (rnr) [inline Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift]: shouldn't we use here `mockConnectivity`?

File: Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+        )
+    }
+    
+    override func tearDown() {
+        mockInteractor = nil
+        viewModel = nil
+        super.tearDown()
+    }
+    
+    private func makeSUT(
```
Comment (rnr) [inline Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift]: where do you use this function?

File: Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+        
+        await viewModel.markNotificationsAsSeen()
+        
+        Verify(mockInteractor, .markNotificationsAsSeen())
+    }
+    
+    func testGetAllNotificationsSuccess() async throws {
+        Given(mockInteractor, .getAllNotifications(page: 1, willReturn: notifications))
+        
+        viewModel = NotificationsInboxViewModel(
```
Comment (rnr) [inline Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift]: Why do we need to create viewModel once again if we did it in setUp?

File: Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+    }
+    
+    func testMarkNotificationAsReadSuccess() async throws {
+        let expectedResponse = NotificationsSeenRead(message: "Notification marked read.")
+        
+        Given(mockInteractor, .markNotificationAsRead(notificationId: .any, willReturn: expectedResponse))
+        
+        await viewModel.markNotificationAsRead(notificationId: "1")
+        
+        Verify(mockInteractor, .markNotificationAsRead(notificationId: .any))
```
Comment (rnr) [inline Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift]: Lets add `count` parameter here (how much times func was called) - and do this for all tests?
Also if we called `markNotificationAsRead` for Notification with id="1" then it makes sense to `Verify` this value but not `.any` (same in another tests).
Also let's check we don't have error with `XCTAssertNil(viewModel.errorMessage)`

File: Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+        
+        XCTAssertNotNil(viewModel.errorMessage)
+    }
+    
+    func testMarkNotificationAsReadUnknownError() async throws {
+        let noInternetError = AFError.sessionInvalidated(error: NSError(domain: "error", code: -1, userInfo: nil))
```
Comment (rnr) [inline Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift]: bad let name for `unknown` error

File: Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+    func testMarkNotificationAsReadNoInternetError() async throws {
+        let noInternetError = AFError.sessionInvalidated(error: URLError(.notConnectedToInternet))
+        
+        Given(mockInteractor, .markNotificationAsRead(notificationId: .any, willThrow: noInternetError))
+        
+        await viewModel.markNotificationAsRead(notificationId: "1")
+        
+        Verify(mockInteractor, .markNotificationAsRead(notificationId: .any))
+        
+        XCTAssertNotNil(viewModel.errorMessage)
```
Comment (rnr) [inline Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift]: if we just check `errorMessage != nil` then this test does the same as next one (testMarkNotificationAsReadUnknownError) because `errorMessage` gets value for any throwing error.
Maybe it's better to check if `errorMessage` got particular error message (`CoreLocalization.Error.slowOrNoInternetConnection` in this test)?
WDYT?

File: Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+        let expectedResponse = NotificationsSeenRead(message: "Notifications marked read.")
+        
+        Given(mockInteractor, .markAllNotificationsAsRead(willReturn: expectedResponse))
+        
+        await viewModel.loadNotifications()
+        await viewModel.markAllNotificationsAsRead()
+        
+        Verify(mockInteractor, .markAllNotificationsAsRead())
+        
+        XCTAssertNotNil(viewModel.flatNotifications[0].lastRead)
```
Comment (rnr) [inline Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift]: Maybe we need to check all notifications, not only one?

Comment (rnr) [reply Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift]: Also will be good to check if errorMessage is `nil`

File: Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+    
+    func testMarkAllNotificationsAsReadNoInternetError() async throws {
+        Given(mockInteractor, .markAllNotificationsAsRead(willThrow: URLError(.notConnectedToInternet)))
+        
+        await viewModel.loadNotifications()
+        await viewModel.markAllNotificationsAsRead()
+        
+        Verify(mockInteractor, .markAllNotificationsAsRead())
+        
+        XCTAssertNotNil(viewModel.errorMessage)
```
Comment (rnr) [inline Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift]: same as for `testMarkNotificationAsReadNoInternetError`

File: Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+        
+        Verify(mockInteractor, .markNotificationsAsSeen())
+    }
+    
+    func testMarkNotificationsAsSeenNoInternetError() async throws {
+        Given(mockInteractor, .markNotificationsAsSeen(willThrow: URLError(.notConnectedToInternet)))
+        
+        await viewModel.markNotificationsAsSeen()
+        
+        Verify(mockInteractor, .markNotificationsAsSeen())
```
Comment (rnr) [inline Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift]: As I see errors are not catching for `markNotificationsAsSeen`. So what we are testing in this and next test?

File: Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+    func testMarkNotificationsAsSeenUnknownError() async throws {
+        Given(mockInteractor, .markNotificationsAsSeen(willThrow: NSError(domain: "error", code: -1, userInfo: nil)))
+        
+        await viewModel.markNotificationsAsSeen()
+        
+        Verify(mockInteractor, .markNotificationsAsSeen())
+    }
+    
+    func testGetAllNotificationsSuccess() async throws {
+        Given(mockInteractor, .getAllNotifications(page: 1, willReturn: notifications))
```
Comment (rnr) [inline Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift]: We did this in setUp already, right?

File: Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift [hunk 12]
```diff
@@ -0,0 +1,309 @@
...
+        await viewModel.loadNotifications()
+        
+        XCTAssertEqual(notifications.results?.count, viewModel.flatNotifications.count)
+    }
+    
+    func testGetAllNotificationsNoInternetError() async throws {
+        Given(mockInteractor, .getAllNotifications(page: 1, willThrow: URLError(.notConnectedToInternet)))
+        Given(mockConnectivity, .isInternetAvaliable(getter: false))
+        
+        viewModel = NotificationsInboxViewModel(
```
Comment (rnr) [inline Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift]: same here

File: Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift [hunk 13]
```diff
@@ -0,0 +1,309 @@
...
+        await viewModel.loadNotifications()
+        
+        XCTAssertEqual(viewModel.screenState, .noInternet)
+    }
+    
+    func testGetAllNotificationsUnknownError() async throws {
+        Given(mockInteractor, .getAllNotifications(page: 1, willThrow: NSError(domain: "error", code: -1, userInfo: nil)))
+        Given(mockConnectivity, .isInternetAvaliable(getter: true))
+        
+        viewModel = NotificationsInboxViewModel(
```
Comment (rnr) [inline Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift]: same here

File: Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift [hunk 14]
```diff
@@ -0,0 +1,309 @@
...
+        await viewModel.loadNotifications()
+        
+        XCTAssertEqual(viewModel.flatNotifications[2].notificationType, "new_response")
+    }
+    
+    func testNotificationTypeAsNewComment() async throws {
+        await viewModel.loadNotifications()
+        
+        XCTAssertEqual(viewModel.flatNotifications[3].notificationType, "new_comment")
+    }
```
Comment (rnr) [inline Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift]: Do you think we need all these tests? Which NotificationsInboxViewModel's function we test here? 
By fact we just check if string from mock let `notifications` is equal to another string.

File: Notifications/Notifications/Presentation/NotificationsInbox/NotificationsInboxViewModel.swift [hunk 15]
```diff
@@ -31,7 +31,7 @@ public class NotificationsInboxViewModel: ObservableObject {
     private var interactor: NotificationsInteractorProtocol
     private var cancellables = Set<AnyCancellable>()
     private(set) var lazyVStackUniqueID: String = UUID().uuidString
-    private var flatNotifications: [SingleNotification] = [] {
+    var flatNotifications: [SingleNotification] = [] {
```
Comment (rnr) [inline Notifications/Notifications/Presentation/NotificationsInbox/NotificationsInboxViewModel.swift]: `private(set)` is better then

**Review by rnr — CHANGES_REQUESTED** [review]

**Review by mta452 — CHANGES_REQUESTED** [review]
Here’s my general feedback on this:
* It would be better to use JSON string directly and decode it into respective data types if possible.
* It would be better to use `makeSUT(...)` for preparing the relevant variables in each test case instead of relying on general `setUp()` method. Also, renaming `viewModel` to either `sut` or `subject` would clearly indicate the core data type being tested.
* Assertions for supporting side effects should also be added in each test case.
* Test cases for menu selection can be added to verify proper api calls, navigation, analytics etc.
* Test cases for load/refresh/fetchMore notifications should be added to verify the proper screen states as well as the notifications data.
* A test case to check the proper grouping of notifications would be really helpful to catch unexpected future bugs around this.
* I remember we struggled with an issue where scrolling to the bottom of the list wasn’t triggering the load more operation. A test case would be very helpful to ensure that the infinite scrolling behavior remains intact.

**Comment by shafqat-muneer** [comment]
> * It would be better to use JSON string directly and decode it into respective data types if possible.

Right now, this approach is being used throughout the code, and we're focusing on testing the `NotificationsInboxViewModel` functionality. We can write test cases for the decoding part later, as part of a separate ticket.

> * It would be better to use `makeSUT(...)` for preparing the relevant variables in each test case instead of relying on general `setUp()` method. Also, renaming `viewModel` to either `sut` or `subject` would clearly indicate the core data type being tested.

- Renamed viewModel to sut.
- Using `makeSUT()` instead of `setUp()` is a great suggestion. However, in our current tests, we’re using the same mocks and setup across all cases, and we're not modifying the `viewModel` dependencies per test. So, at this stage, using `setUp()` seems appropriate. We can revisit this and switch to `makeSUT()` later if we need to override behavior in specific test cases.

> * Assertions for supporting side effects should also be added in each test case.

If you're referring to `XCTAssert` comparisons, I've added a few more of them.

> * Test cases for menu selection can be added to verify proper api calls, navigation, analytics etc.
> * Test cases for load/refresh/fetchMore notifications should be added to verify the proper screen states as well as the notifications data.
> * A test case to check the proper grouping of notifications would be really helpful to catch unexpected future bugs around this.
> * I remember we struggled with an issue where scrolling to the bottom of the list wasn’t triggering the load more operation. A test case would be very helpful to ensure that the infinite scrolling behavior remains intact.

These would be a nice addition to the test cases. Our intention was to get started with some basic tests and gradually build on them. We can enhance and add more test cases as part of a separate ticket.

File: Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift (hunk 3 above)
Comment (shafqat-muneer) [reply Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift]: It was written to use it but later there were not any case. Just removed for now. Thanks for highlighting.

File: Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift (hunk 4 above)
Comment (shafqat-muneer) [reply Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift]: Intention behind was to execute full functionality at a single test case level but you are also right. For now, I just removed re-initialization from single test case level.

File: Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift (hunk 10 above)
Comment (shafqat-muneer) [reply Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift]: I think, it's a valid and meaningful test. We’re not testing the result or state, but we are verifying system behavior that the interactor is called even under error conditions — and the system remains stable.

File: Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift (hunk 12 above)
Comment (shafqat-muneer) [reply Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift]: https://github.com/edx/edx-mobile-marketplace-ios/pull/148#discussion_r2140776025

File: Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift (hunk 13 above)
Comment (shafqat-muneer) [reply Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift]: https://github.com/edx/edx-mobile-marketplace-ios/pull/148#discussion_r2140776025

**Comment by shafqat-muneer** [comment]
@rnr @mta452 I've addressed and commented on the review feedback. The PR is ready for another round of review. Thanks!

File: Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift [hunk 16]
```diff
@@ -0,0 +1,259 @@
...
+    func testMarkNotificationAsReadNoInternetError() async throws {
+        let noInternetError = AFError.sessionInvalidated(error: URLError(.notConnectedToInternet))
+        
+        Given(mockInteractor, .markNotificationAsRead(notificationId: .value("1"), willThrow: noInternetError))
+        
+        await sut.markNotificationAsRead(notificationId: "1")
+        
+        Verify(mockInteractor, 1, .markNotificationAsRead(notificationId: .value("1")))
+        
+        XCTAssertNotNil(sut.errorMessage)
```
Comment (rnr) [inline Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift:140]: next line will fail if errorMessage is `nil`. Do we need both?

File: Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift [hunk 17]
```diff
@@ -0,0 +1,259 @@
...
+        let noInternetError = AFError.sessionInvalidated(error: URLError(.notConnectedToInternet))
+        
+        Given(mockInteractor, .markAllNotificationsAsRead(willThrow: noInternetError))
+        
+        await sut.loadNotifications()
+        await sut.markAllNotificationsAsRead()
+        
+        Verify(mockInteractor, 1, .markAllNotificationsAsRead())
+        
+        XCTAssertNotNil(sut.errorMessage)
```
Comment (rnr) [inline Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift:188]: the same: next line will fail if errorMessage is `nil`

File: Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift [hunk 18]
```diff
@@ -0,0 +1,259 @@
...
+    
+    func testMarkAllNotificationsAsReadNoInternetError() async throws {
+        let noInternetError = AFError.sessionInvalidated(error: URLError(.notConnectedToInternet))
+        
+        Given(mockInteractor, .markAllNotificationsAsRead(willThrow: noInternetError))
+        
+        await sut.loadNotifications()
+        await sut.markAllNotificationsAsRead()
+        
+        Verify(mockInteractor, 1, .markAllNotificationsAsRead())
```
Comment (rnr) [inline Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift:182]: maybe we need to check `lastRead` is still `nil`? like `XCTAssertNil(sut.flatNotifications[0].lastRead)`

File: Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift [hunk 19]
```diff
@@ -0,0 +1,259 @@
...
+    
+    func testMarkAllNotificationsAsReadUnknownError() async throws {
+        let unknown = AFError.sessionInvalidated(error: NSError(domain: "error", code: -1, userInfo: nil))
+        
+        Given(mockInteractor, .markAllNotificationsAsRead(willThrow: unknown))
+        
+        await sut.loadNotifications()
+        await sut.markAllNotificationsAsRead()
+        
+        Verify(mockInteractor, 1, .markAllNotificationsAsRead())
```
Comment (rnr) [inline Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift:200]: same `XCTAssertNil(sut.flatNotifications[0].lastRead)` here?

File: Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift (hunk 10 above)
Comment (rnr) [reply Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift]: ok, but why we need to do it twice (for `no internet` and `unknown` errors) if it doesn't depend on error type?

File: Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift [hunk 20]
```diff
@@ -0,0 +1,259 @@
...
+        
+        await sut.markNotificationsAsSeen()
+        
+        Verify(mockInteractor, 1, .markNotificationsAsSeen())
+    }
+    
+    func testGetAllNotificationsSuccess() async throws {
+        await sut.loadNotifications()
+        
+        XCTAssertEqual(notifications.results?.count, sut.flatNotifications.count)
```
Comment (rnr) [inline Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift:234]: maybe add verification of calls count for `.getAllNotifications`?

File: Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift [hunk 21]
```diff
@@ -0,0 +1,259 @@
...
+    
+    func testGetAllNotificationsNoInternetError() async throws {
+        let noInternetError = AFError.sessionInvalidated(error: URLError(.notConnectedToInternet))
+        
+        Given(mockInteractor, .getAllNotifications(page: 1, willThrow: noInternetError))
+        Given(mockConnectivity, .isInternetAvaliable(getter: false))
+        
+        await sut.loadNotifications()
+        
+        XCTAssertEqual(sut.screenState, .noInternet)
```
Comment (rnr) [inline Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift:249]: check `sut.flatNotifications.count` is equal `zero` here?
and verification of calls count for .getAllNotifications?

File: Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift [hunk 22]
```diff
@@ -0,0 +1,259 @@
...
+    
+    func testGetAllNotificationsUnknownError() async throws {
+        let unknown = AFError.sessionInvalidated(error: NSError(domain: "error", code: -1, userInfo: nil))
+        
+        Given(mockInteractor, .getAllNotifications(page: 1, willThrow: unknown))
+        Given(mockConnectivity, .isInternetAvaliable(getter: true))
+        
+        await sut.loadNotifications()
+        
+        XCTAssertEqual(sut.screenState, .serverError)
```
Comment (rnr) [inline Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift:263]: same here

**Review by rnr — CHANGES_REQUESTED** [review]
Thanks for the good work. Just a few more questions )

File: Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift (hunk 16 above)
Comment (shafqat-muneer) [reply Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift:140]: I’ve added the first assertion to explicitly confirm that errorMessage is set, and the second to verify it matches `.slowOrNoInternetConnection`. This helps isolate whether the failure is due to a missing error or an unexpected value, making the test output clearer. Let me know if you’d prefer simplifying it.

File: Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift (hunk 17 above)
Comment (shafqat-muneer) [reply Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift:188]: https://github.com/edx/edx-mobile-marketplace-ios/pull/148#discussion_r2149787911

File: Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift (hunk 18 above)
Comment (shafqat-muneer) [reply Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift:182]: Added against mock elements.

File: Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift (hunk 19 above)
Comment (shafqat-muneer) [reply Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift:200]: Added against mock elements.

File: Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift (hunk 10 above)
Comment (shafqat-muneer) [reply Notifications/NotificationsTests/Presentation/NotificationsInboxViewModelTests.swift]: Makes sense — since the behavior is the same for both errors, I’ve remove one of the test cases to avoid redundancy.

**Comment by shafqat-muneer** [comment]
@rnr I've addressed and commented on the review feedback. Kindly have another round of review. Thanks!

**Review by rnr — APPROVED** [review]
LGTM. Thank you

**Review by mta452 — APPROVED** [review]
The implemented test cases look good, would be best if suggested test cases are also added.

**Comment by shafqat-muneer** [comment]
> The implemented test cases look good, would be best if suggested test cases are also added.

These would be a nice addition to the test cases. Our intention was to get started with some basic tests and gradually build on them. We can enhance and add more test cases as part of a separate ticket.
